### PR TITLE
[SPARK-43274][SPARK-43275][PYTHON][CONNECT] Introduce `PySparkNotImplementedError`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to introduce `PySparkNotImplementedError`. This PR also includes migrating Spark Connect GroupedData errors into error class as well.

### Why are the changes needed?

To cover user-facing Python built-in error by PySpark error framework for improving error message usability.

### Does this PR introduce _any_ user-facing change?

No API changes.

### How was this patch tested?

Fixed UT, the existing CI should pass.

### CODE REVIEW

File: python/pyspark/errors/error_classes.py [hunk 1]
```diff
@@ -269,6 +269,11 @@
       "<objectName> is not iterable."
     ]
   },
+  "NOT_LIST" : {
```
Comment (amaliujia) [inline python/pyspark/errors/error_classes.py:272]: We should consolidate these `NOT_xxx` to one error as they all throw `xx should be but got yy`?

Comment (itholic) [reply python/pyspark/errors/error_classes.py:272]: Yes, we can classify error classes starting with `NOT_XXX` into one error class or group similar types of errors into a main-sub error class structure. Each way has its pros and cons, so we plan to proceed with the initial migration in the current form first, and then gather opinions from ppl through decision docs or something before making changes all at once.

BTW, we already have a [ticket](https://issues.apache.org/jira/browse/SPARK-43020) for this FYI :-)

Comment (amaliujia) [reply python/pyspark/errors/error_classes.py:272]: Sounds good then!

**Review by amaliujia — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**